atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,308 @@
-        data = self._ssh_run_remote_command(cmd)
-        return data
+    def __init__(self, *, hostname: str, username: str, ssh_keyfile: str):
+        """
+        :param hostname: str, remote hostname for ssh connection
+        :param username: str, username for ssh connection
+        :param ssh_keyfile: str, filepath to ssh private key
+        """
+        self._hostname = hostname
+        self._username = username
+        self._ssh_keyfile = ssh_keyfile
 
-    def _read_slot_data(self, slot: int):
-        cmd = f"cat {ROOT_PATH}slot_{slot}.json"
-        data = self._ssh_run_remote_command(cmd)
-        return data
+    @property
+    def hostname(self):
+        return self._hostname
+
+    @property
+    def username(self):
+        return self._username
+
+    @property
+    def ssh_keyfile(self):
+        return self._ssh_keyfile
 
     def _validate_resource_lock(self, payload: dict):
+        """ Validate resource lock json file against schema """
         with open(RESOURCE_LOCK_SCHEMA, 'r') as f:
             schema_data = f.read()
         schema = json.loads(schema_data)
         try:
             jsonschema.validate(payload, schema)
-            return True
         except jsonschema.ValidationError as err:
-            logger.error(err)
-        return False
+            logger.warning("Invalid json file: %s", err)
+
+    @contextmanager
+    def _ssh_session(self):
+        """ Create an SSH connection."""
+        client = self._open_ssh_session()
+        yield client
+        client.close()
+
+    @backoff.on_exception(
+        backoff.expo,
+        RETRY_ON_SSH_EXCEPTIONS,
+        factor=BACKOFF_FACTOR,
+        max_tries=MAX_RETRIES,
+        jitter=None,  # use deterministic backoff, do not apply random jitter
+        logger=logger,
+    )
+    def _open_ssh_session(self):
+        """
+        Create a new SSH connection and return connection object.
+        """
+        client = paramiko.SSHClient()
+        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        logger.debug("Open SSH connection in remote host %s", self.hostname)
+        client.connect(
+            self.hostname, username=self.username, key_filename=self.ssh_keyfile)
+        return client
+
+    @backoff.on_exception(
+        backoff.expo,
+        RETRY_ON_SSH_EXCEPTIONS,
+        factor=BACKOFF_FACTOR,
+        max_tries=MAX_RETRIES,
+        jitter=None,  # use deterministic backoff, do not apply random jitter
+        logger=logger,
+    )
+    def _ssh_run_remote_cmd(self, client, cmd: List):
+        """
+        Run a Shell command on a remote host
 
-    def check_avail(self):
+        :return: stdout, stderr and exitcode of shell command
+        """
+        logger.debug("Try to run command %s on remote host %s", cmd, self.hostname)
+        _, stdout, stderr = client.exec_command(cmd, timeout=SSH_COMMAND_TIMEOUT)
+        return stdout, stderr, client.recv_exit_status()
```

Comment:
```suggestion
        return stdout, stderr, stdout.channel.recv_exit_status()
```
The Client class does not have this method.

Also, according to [Paramiko docs](https://docs.paramiko.org/en/stable/api/channel.html#paramiko.channel.Channel.recv_exit_status), `recv_exit_status` may hang indefinitely unless you read from the output streams first. Maybe this method should fully consume stdout and stderr, and return strings rather than the open Channel Files.